src/CopyOnWrite/Copy.cs

```diff
@@ -0,0 +1,1044 @@
+        }
+
+        /// <summary>
+        /// Try to get a message to inform the user which processes have a lock on a given file.
+        /// </summary>
+        private static string GetLockedFileMessage(string file)
+        {
+            string message = string.Empty;
+
+            try
+            {
+                var processes = LockCheck.GetProcessesLockingFile(file);
+                message = !string.IsNullOrEmpty(processes)
+                    ? string.Format("The file is locked by: \"{0}\"", processes)
+                    : String.Empty;
+            }
+            catch (Exception)
+            {
+                // Never throw if we can't get the processes locking the file.
+            }
+
+            return message;
+        }
+
+        /// <summary>
+        /// Standard entry point.
+        /// </summary>
+        /// <returns></returns>
+        public override bool Execute()
+        {
+            return Execute(CopyFileWithLogging, s_parallelism);
+        }
+
+        #endregion
+
+        /// <summary>
+        /// Compares two paths to see if they refer to the same file. We can't solve the general
+        /// canonicalization problem, so we just compare strings on the full paths.
+        /// </summary>
+        private static bool PathsAreIdentical(string source, string destination)
+        {
+            string fullSourcePath = Path.GetFullPath(source);
+            string fullDestinationPath = Path.GetFullPath(destination);
+            StringComparison filenameComparison = NativeMethods.IsWindows ? StringComparison.OrdinalIgnoreCase : StringComparison.Ordinal;
+            return String.Equals(fullSourcePath, fullDestinationPath, filenameComparison);
+        }
+
+        private static int GetParallelismFromEnvironment()
+        {
+            int parallelism = CopyTaskParallelism;
+            if (parallelism < 0)
+            {
+                parallelism = DefaultCopyParallelism;
+            }
+            else if (parallelism == 0)
+            {
+                parallelism = int.MaxValue;
+            }
+            return parallelism;
+        }
+
+        /// <summary>
+        /// Setting the associated environment variable to 1 restores the pre-15.8 single
+        /// threaded (slower) copy behavior. Zero implies Int32.MaxValue, less than zero
+        /// (default) uses the empirical default in Copy.cs, greater than zero can allow
+        /// perf tuning beyond the defaults chosen.
+        /// </summary>
+        public static readonly int CopyTaskParallelism = ParseIntFromEnvironmentVariableOrDefault("MSBUILDCOPYTASKPARALLELISM", -1);
+
+        private static int ParseIntFromEnvironmentVariableOrDefault(string environmentVariable, int defaultValue)
+        {
+            return int.TryParse(Environment.GetEnvironmentVariable(environmentVariable), out int result)
+                ? result
+                : defaultValue;
+        }
+
+        #region CopyOnWrite functionality
```

Comment:
This is the majority of my change to the Copy task (till the end of file).

Reply:
Hmm in a perfect world, we would update the stock `Copy` task to allow us to override it.  But then none of this would work with older versions of MSBuild 😢 Are you going to keep this up-to-date when the `Copy` task changes? 😄